File: backend/agent/materials.py
"""v0.19 零件材质推断（后端侧，零 CAD 依赖）。

权威来源是内核 `mech_kernel/materials.py`。后端不能 import 内核（内核依赖
OCC / build123d），所以这里复刻同一套「零件名 → 材质键 + 基色」规则，并由
`tests/test_material_consistency.py` 按路径加载内核文件做一致性断言，防止漂移。

归档零件时用它把 material / material_color 写进项目零件库 manifest，
前端视口据此按材质着色（工程 CAD 风格的哑光分色）。
"""

from __future__ import annotations

# 与内核 MATERIALS 的 base 色（0..1）一致。改这里必须同步内核并跑一致性测试。
MATERIAL_BASE: dict[str, tuple[float, float, float]] = {
    "steel": (0.55, 0.62, 0.72),
    "cast_iron": (0.47, 0.49, 0.51),
    "aluminum": (0.70, 0.73, 0.76),
    "bronze": (0.76, 0.55, 0.25),
    "copper": (0.72, 0.45, 0.30),
    "brass": (0.74, 0.62, 0.32),
    "titanium": (0.55, 0.55, 0.58),
    "dark_steel": (0.36, 0.38, 0.42),
    "nylon": (0.86, 0.86, 0.82),
    "rubber": (0.16, 0.16, 0.17),
}

# 与内核 _NAME_HINTS 同序、同关键词（顺序有意义：拨叉须先于「轴」匹配）。
_NAME_HINTS: tuple[tuple[tuple[str, ...], str], ...] = (
    (("箱体", "箱盖", "壳", "盖", "housing", "case", "cover"), "cast_iron"),
    (("拨叉", "fork"), "bronze"),
    (("同步器", "接合套", "synchron"), "dark_steel"),
    (("轴承", "bearing", "铜", "bronze", "衬套", "bush"), "bronze"),
    (("齿轮", "gear", "惰轮", "idler", "轴", "shaft", "花键", "spline"), "steel"),
)


def resolve_material(name: str) -> str:
    """按零件名推断材质键；未命中返回 steel（与内核一致）。"""
    low = str(name).lower()
    for keys, mat in _NAME_HINTS:
        if any(k in low for k in keys):
            return mat
    return "steel"


def material_color(name: str) -> list[float]:
    """按零件名推断材质基色（0..1 RGB，manifest 序列化用 list）。"""
    base = MATERIAL_BASE.get(resolve_material(name), MATERIAL_BASE["steel"])
    return [float(c) for c in base]

File: frontend/src/materials.ts
/**
 * v0.19 工程 CAD 视口材质表。
 *
 * 与内核 `mech_kernel/materials.py` 的 MATERIALS 基色对齐（权威来源在内核）；
 * 前端只做展示，用非金属哑光着色（metalness=0）——这是 CAD 视口的惯例：
 * 分材质区分零件，但不做写实金属反射，保持工程图的均匀可读性。
 *
 * 键名与内核 resolve_material() 的返回值一致；后端归档时会把 material 与
 * material_color 写进 manifest，前端优先用后端的解析结果，本地表仅作回退。
 */
export type CadMaterial = {
  color: number;
  roughness: number;
  metalness: number;
};

// 基色取内核 materials.py 的 base（0..1）转 hex；roughness 按材质给人的
// 粗糙感设定（铸铁/橡胶粗糙，青铜/黄铜稍亮），metalness 统一 0。
export const MATERIAL_CAD: Record<string, CadMaterial> = {
  steel: { color: 0x8c9eb8, roughness: 0.42, metalness: 0.0 },
  cast_iron: { color: 0x787c82, roughness: 0.72, metalness: 0.0 },
  aluminum: { color: 0xb3bac2, roughness: 0.38, metalness: 0.0 },
  bronze: { color: 0xc28c40, roughness: 0.40, metalness: 0.0 },
  copper: { color: 0xb8734d, roughness: 0.40, metalness: 0.0 },
  brass: { color: 0xbd9e52, roughness: 0.38, metalness: 0.0 },
  titanium: { color: 0x8c8c94, roughness: 0.46, metalness: 0.0 },
  dark_steel: { color: 0x5c616b, roughness: 0.50, metalness: 0.0 },
  nylon: { color: 0xdbdbd1, roughness: 0.78, metalness: 0.0 },
  rubber: { color: 0x29292b, roughness: 0.92, metalness: 0.0 },
};

/** 未命中材质键时的中性 CAD 灰。 */
export const CAD_DEFAULT: CadMaterial = { color: 0x9aa7b6, roughness: 0.5, metalness: 0.0 };

/** 材质键 → CAD 材质；未知键回退中性灰。 */
export function resolveCadMaterial(key?: string | null): CadMaterial {
  if (key && MATERIAL_CAD[key]) {
    return MATERIAL_CAD[key];
  }
  return CAD_DEFAULT;
}

/**
 * 内核 material_color（0..1 浮点）→ three 用的 hex。
 * 后端 manifest 里的 material_color 是权威色；无效值回退到材质表/中性灰。
 */
export function colorFromRgb(rgb?: number[] | null, fallbackKey?: string | null): number {
  if (Array.isArray(rgb) && rgb.length >= 3 && rgb.every((v) => Number.isFinite(v))) {
    const clamp = (v: number) => Math.max(0, Math.min(255, Math.round(v * 255)));
    return (clamp(rgb[0]) << 16) | (clamp(rgb[1]) << 8) | clamp(rgb[2]);
  }
  return resolveCadMaterial(fallbackKey).color;
}

/** 按零件名关键词推断材质（与内核 _NAME_HINTS 同序，仅在前端缺后端字段时用）。 */
export function resolveCadMaterialFromName(name: string): string {
  const low = String(name || "").toLowerCase();
  const hints: [string[], string][] = [
    [["箱体", "箱盖", "壳", "盖", "housing", "case", "cover"], "cast_iron"],
    [["拨叉", "fork"], "bronze"],
    [["同步器", "接合套", "synchron"], "dark_steel"],
    [["轴承", "bearing", "铜", "bronze", "衬套", "bush"], "bronze"],
    [["齿轮", "gear", "惰轮", "idler", "轴", "shaft", "花键", "spline"], "steel"],
  ];
  for (const [keys, mat] of hints) {
    if (keys.some((k) => low.includes(k))) {
      return mat;
    }
  }
  return "steel";
}
